### PR TITLE
Introduce inference ids for quantifier instantiation

### PR DESCRIPTION
**Pull request by ajreynol**
This makes quantifiers use standard inference ids.

This eliminates the need to track ad-hoc statistics, per instantiation type.

This makes minor modifications to a few interfaces in quantifiers to enable this.

### CODE REVIEW

File: src/theory/quantifiers/ematching/ho_trigger.cpp [hunk 1]
```diff
@@ -372,27 +372,29 @@ bool HigherOrderTrigger::sendInstantiation(InstMatch& m)
...
-    return d_quantEngine->getInstantiate()->addInstantiation(d_quant, m.d_vals);
+    return d_quantEngine->getInstantiate()->addInstantiation(d_quant, m, id);
   }
 }
 
 // recursion depth limited by number of arguments of higher order variables
 // occurring as pattern operators (very small)
-bool HigherOrderTrigger::sendInstantiation(InstMatch& m, unsigned var_index)
+bool HigherOrderTrigger::sendInstantiation(std::vector<Node>& m,
+                                           unsigned var_index)
```
Comment (nafur) [inline src/theory/quantifiers/ematching/ho_trigger.cpp:382]:
```suggestion
                                           uint32_t var_index)
```

File: src/theory/quantifiers/ematching/ho_trigger.h [hunk 2]
```diff
@@ -235,7 +235,7 @@ class HigherOrderTrigger : public Trigger
   *   when var_index = 0,1, we are processing possibilities for
   *    instantiation of f1,f2 respectively.
   */
-  bool sendInstantiation(InstMatch& m, unsigned var_index);
+  bool sendInstantiation(std::vector<Node>& m, unsigned var_index);
```
Comment (nafur) [inline src/theory/quantifiers/ematching/ho_trigger.h:238]:
```suggestion
  bool sendInstantiation(std::vector<Node>& m, uint32_t var_index);
```

File: src/theory/quantifiers/ematching/inst_match_generator.cpp [hunk 3]
```diff
@@ -559,7 +563,7 @@ uint64_t InstMatchGenerator::addInstantiations(Node f,
   while (getNextMatch(f, m, qe, tparent) > 0)
   {
     if( !d_active_add ){
-      if (sendInstantiation(tparent, m))
+      if (sendInstantiation(tparent, m, InferenceId::UNKNOWN))
```
Comment (nafur) [inline src/theory/quantifiers/ematching/inst_match_generator.cpp:566]: What about this UNKNOWN?

**Review by nafur — APPROVED** [review]
LGTM, just some minor comments.

File: src/theory/quantifiers/ematching/inst_match_generator.cpp (hunk 3 above)
Comment (ajreynol) [reply src/theory/quantifiers/ematching/inst_match_generator.cpp:566]: This is dead code that I believe will be deleted at some point.

File: src/theory/quantifiers/ematching/ho_trigger.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/quantifiers/ematching/ho_trigger.cpp:382]: I'm going to change to `size_t`.

**Pull request merged**